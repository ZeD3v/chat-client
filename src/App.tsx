import React, { useState, Dispatch, SetStateAction, useMemo } from 'react';

import Login from './pages/Login';
import Chat from './pages/Chat';

import styles from './common/styles/App.module.scss';

export type User = {
  username: string;
  token: string;
};

type UserContext = {
  user: User | null;
  setUser: Dispatch<SetStateAction<User | null>> | null;
};

type ErrorContext = {
  error: string | null;
  setError: Dispatch<SetStateAction<string | null>> | null;
};

export const UserContext = React.createContext<UserContext>({
  user: null,
  setUser: null,
});

export const ErrorContext = React.createContext<ErrorContext>({
  error: null,
  setError: null,
});

const App: React.FC = () => {
  const [user, setUser] = useState<User | null>(null);
  const [error, setError] = useState<string | null>(null);

  const userContextValue = useMemo(() => ({ user, setUser }), [user, setUser]);
  const errorContextValue = useMemo(() => ({ error, setError }), [
    error,
    setError,
  ]);

  return (
    <UserContext.Provider value={userContextValue}>
      <ErrorContext.Provider value={errorContextValue}>
        <div className={styles.container}>
          <div className={styles['inner-container']}>
            {/*
          For this simple scenario, basic if block will do.
          When page count grows and url navigation support becomes
          necessary - react-router or some similar navigation solution
          should be introduced.
        */}
            {(!user && <Login />) || <Chat />}
          </div>
        </div>
      </ErrorContext.Provider>
    </UserContext.Provider>
  );
};

export default App;
