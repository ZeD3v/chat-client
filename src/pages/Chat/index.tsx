import React, {
  useState,
  useRef,
  useEffect,
  useContext,
  useCallback,
} from 'react';
import io from 'socket.io-client';

import Spinner from '../../common/components/Spinner';

import { UserContext, ErrorContext } from '../../App';

import UserConnectedEvent from './Events/UserConnectedEvent';
import MessageEvent from './Events/MessageEvent';
import UserDisconnectedEvent from './Events/UserDisconnectedEvent';

import buttonStyles from '../../common/styles/Button.module.scss';
import styles from './styles/Chat.module.scss';

export enum ChatEvents {
  'MESSAGE' = 'MESSAGE',
  'USER_CONNECTED' = 'USER_CONNECTED',
  'USER_DISCONNECTED' = 'USER_DISCONNECTED',
}

type MessageSocketData = {
  eventId: string;
  username: string;
  message: string;
  timestamp: number;
};

type UserConnectedSocketData = {
  eventId: string;
  username: string;
  timestamp: number;
};

type UserDisconnectedSocketData = {
  eventId: string;
  username: string;
  reason: string;
  timestamp: number;
};

type ChatEvent = UserConnectedEvent | MessageEvent | UserDisconnectedEvent;

const ChatPage: React.FC = () => {
  const messagesContainerRef = useRef<HTMLDivElement>(null);
  const socketRef = useRef<SocketIOClient.Socket>();

  const { user, setUser } = useContext(UserContext);
  const { setError } = useContext(ErrorContext);

  const [events, setEvents] = useState<ChatEvent[]>([]);
  const [isLoading, setIsLoading] = useState(true);
  const [newMessage, setNewMessage] = useState<string>('');

  const sendMessage = (): void => {
    if (socketRef.current) {
      socketRef.current.emit('message', {
        message: newMessage,
      });
    }
  };

  const prepareToDisconnect = useCallback(
    (reason?: string): void => {
      setEvents([]);
      setNewMessage('');

      if (reason && setError) {
        setError(reason);
      }

      if (setUser) {
        setUser(null);
      }
    },
    [setError, setUser]
  );

  const onSendMessage = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();

    sendMessage();
    setNewMessage('');
  };

  useEffect(() => {
    socketRef.current = io(process.env.REACT_APP_API_URL || '/', {
      query: `${user ? `username=${user.username}&token=${user.token}` : ''}`,
    });

    socketRef.current.on('connect', (): void => {
      // To avoid flash of loading spinner and imporve ux
      // show loading spinner 0.5s longer after the actual
      // connection gets established
      setTimeout(() => {
        setIsLoading(false);
      }, 500);
    });

    socketRef.current.on('error', (): void => {
      prepareToDisconnect(
        'Oops. Something went wrong, please try again later.'
      );
    });

    socketRef.current.on('message', (data: MessageSocketData): void => {
      setEvents((currentEvents) => [
        ...currentEvents,
        {
          type: ChatEvents.MESSAGE,
          id: data.eventId,
          timestamp: data.timestamp,
          payload: {
            username: data.username,
            message: data.message,
          },
        },
      ]);
    });

    socketRef.current.on(
      'user_connected',
      (data: UserConnectedSocketData): void => {
        setEvents((currentEvents) => [
          ...currentEvents,
          {
            type: ChatEvents.USER_CONNECTED,
            id: data.eventId,
            timestamp: data.timestamp,
            payload: {
              username: data.username,
            },
          },
        ]);
      }
    );

    socketRef.current.on(
      'user_disconnected',
      (data: UserDisconnectedSocketData): void => {
        if (
          data.username === user?.username &&
          data.reason === 'USER_INACTIVE'
        ) {
          prepareToDisconnect('Disconnected due to inactivity.');
          return;
        }

        setEvents((currentEvents) => [
          ...currentEvents,
          {
            type: ChatEvents.USER_DISCONNECTED,
            id: data.eventId,
            timestamp: data.timestamp,
            payload: {
              username: data.username,
              reason: data.reason,
            },
          },
        ]);
      }
    );

    socketRef.current.on('server_shutting_down', (): void => {
      prepareToDisconnect('Disconnected, server shutting down.');
    });

    socketRef.current.on('disconnect', (): void => {
      prepareToDisconnect();
    });

    return (): void => {
      if (socketRef.current) {
        socketRef.current.removeAllListeners();
        socketRef.current.disconnect();
      }
    };
  }, [prepareToDisconnect, setError, setUser, user]);

  useEffect(() => {
    if (
      messagesContainerRef &&
      messagesContainerRef.current &&
      messagesContainerRef.current.lastChild
    ) {
      (messagesContainerRef.current.lastChild as HTMLDivElement).scrollIntoView(
        {
          behavior: 'smooth',
          block: 'end',
        }
      );
    }
  }, [events]);

  const renderEvent = (event: ChatEvent): React.ReactNode => {
    if (event.type === ChatEvents.MESSAGE) {
      return <MessageEvent key={event.id} user={user} event={event} />;
    }

    if (event.type === ChatEvents.USER_CONNECTED) {
      return <UserConnectedEvent key={event.id} user={user} event={event} />;
    }

    if (event.type === ChatEvents.USER_DISCONNECTED) {
      return <UserDisconnectedEvent key={event.id} user={user} event={event} />;
    }

    return null;
  };

  return (
    <div className={styles.container}>
      <div className={styles.header}>
        <div>{user?.username}</div>
        <button
          className={`${buttonStyles.button} ${buttonStyles['button--simple']}`}
          type="button"
          onClick={(): void => {
            prepareToDisconnect();
          }}
        >
          Disconnect
        </button>
      </div>
      <div ref={messagesContainerRef} className={styles['messages-container']}>
        {events.map(renderEvent)}
      </div>
      <div className={styles['form-container']}>
        <form onSubmit={onSendMessage} className={styles.form}>
          <textarea
            data-testid="chat-message-input"
            className={styles['message-input']}
            value={newMessage}
            onChange={(event: React.ChangeEvent<HTMLTextAreaElement>): void => {
              setNewMessage(event.target.value);
            }}
          />
          <button
            data-testid="chat-message-submit-btn"
            className={`${buttonStyles.button} ${styles['submit-button']}`}
            type="submit"
          >
            Send!
          </button>
        </form>
        <Spinner
          hidden={!isLoading}
          overlay
          overlayColor="rgb(236, 236, 236)"
        />
      </div>
    </div>
  );
};

export default ChatPage;
