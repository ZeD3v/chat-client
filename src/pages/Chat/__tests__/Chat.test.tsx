import React from 'react';
import { render, fireEvent, getByTestId } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import socketIOClient from 'socket.io-client';

import { UserContext } from '../../../App';
import Chat from '../index';

jest.mock('socket.io-client');

describe('Chat', () => {
  let socket = {
    on: jest.fn(),
    emit: jest.fn(),
    disconnect: jest.fn(),
    removeAllListeners: jest.fn(),
  };

  beforeEach(() => {
    socket = {
      on: jest.fn(),
      emit: jest.fn(),
      disconnect: jest.fn(),
      removeAllListeners: jest.fn(),
    };

    (socketIOClient as any).mockReturnValue(socket);
  });

  test('renders correctly', () => {
    const { asFragment } = render(<Chat />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('sends message when send button is pressed', () => {
    const { container } = render(
      <UserContext.Provider
        value={{
          user: { username: 'Test user', token: 'test-token' },
          setUser: jest.fn(),
        }}
      >
        <Chat />
      </UserContext.Provider>
    );
    const testMessage = 'This is a test message.';

    fireEvent.change(getByTestId(container, 'chat-message-input'), {
      target: { value: testMessage },
    });
    fireEvent.click(getByTestId(container, 'chat-message-submit-btn'));

    expect(socket.emit).toHaveBeenCalledTimes(1);
    expect(socket.emit).toHaveBeenCalledWith('message', {
      message: testMessage,
    });
  });
});
