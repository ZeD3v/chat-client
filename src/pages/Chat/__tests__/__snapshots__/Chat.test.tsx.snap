// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat renders correctly 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="header"
    >
      <div />
      <button
        class="button button--simple"
        type="button"
      >
        Disconnect
      </button>
    </div>
    <div
      class="messages-container"
    />
    <div
      class="form-container"
    >
      <form
        class="form"
      >
        <textarea
          class="message-input"
          data-testid="chat-message-input"
        />
        <button
          class="button submit-button"
          data-testid="chat-message-submit-btn"
          type="submit"
        >
          Send!
        </button>
      </form>
      <div
        class="container container--overlay"
        style="background: rgb(236, 236, 236);"
      >
        <div
          class="inner-container"
        >
          <div
            class="sphere"
          />
          <div
            class="sphere sphere--2"
          />
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
